srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+	/// This will emit the `Transferred` event.
+	pub fn make_transfer_with_event(
+		asset_id: &T::AssetId,
+		from: &T::AccountId,
+		to: &T::AccountId,
+		amount: T::Balance,
+	) -> Result {
+		Self::make_transfer(asset_id, from, to, amount)?;
+
+		if from != to {
+			Self::deposit_event(RawEvent::Transferred(*asset_id, from.clone(), to.clone(), amount));
+		}
+
+		Ok(())
+	}
+
+	/// Move `amount` from free balance to reserved balance.
+	///
+	/// If the free balance is lower than `amount`, then no funds will be moved and an `Err` will
+	/// be returned. This is different behavior than `unreserve`.
+	pub fn reserve(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Result {
+		// Do we need to consider that this is an atomic transaction?
+		let original_reserve_balance = Self::reserved_balance(asset_id, who);
+		let original_free_balance = Self::free_balance(asset_id, who);
+		if original_free_balance < amount {
+			return Err("not enough free funds");
+		}
+		let new_reserve_balance = original_reserve_balance + amount;
+		Self::set_reserved_balance(asset_id, who, new_reserve_balance);
+		let new_free_balance = original_free_balance - amount;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		Ok(())
+	}
+
+	/// Moves up to `amount` from reserved balance to free balance. This function cannot fail.
+	///
+	/// As many assets up to `amount` will be moved as possible. If the reserve balance of `who`
+	/// is less than `amount`, then the remaining amount will be returned.
+	/// NOTE: This is different behavior than `reserve`.
+	pub fn unreserve(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> T::Balance {
+		let b = Self::reserved_balance(asset_id, who);
+		let actual = rstd::cmp::min(b, amount);
+		let original_free_balance = Self::free_balance(asset_id, who);
+		let new_free_balance = original_free_balance + actual;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		Self::set_reserved_balance(asset_id, who, b - actual);
+		amount - actual
+	}
+
+	/// Deduct up to `amount` from the combined balance of `who`, preferring to deduct from the
+	/// free balance. This function cannot fail.
+	///
+	/// As much funds up to `amount` will be deducted as possible. If this is less than `amount`
+	/// then `Some(remaining)` will be returned. Full completion is given by `None`.
+	pub fn slash(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Option<T::Balance> {
+		let free_balance = Self::free_balance(asset_id, who);
+		let free_slash = rstd::cmp::min(free_balance, amount);
+		let new_free_balance = free_balance - free_slash;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		if free_slash < amount {
+			Self::slash_reserved(asset_id, who, amount - free_slash)
+		} else {
+			None
+		}
+	}
+
+	/// Adds up to `amount` to the free balance of `who`.
+	pub fn reward(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Result {
```

Comment:
does this need to be here? are there not adequate functions in the `Currency` trait for it?

Reply:
This one takes an additional parameter `AssetId`.
We are planning to make a `MultiCurrency` trait that similar to `Currency` but takes an additional `AssetId` parameter. The hard part is deal with all the imbalances.

Reply:
something basically useless, like: 
```
MultipleImbalances ( vec! [
  PositiveImbalance { asset_id: 1, value: 200 },
  NegativeImbalance { asset_id: 2, value: 200 },
] )
```

Reply:
as such, i don't think `MultiCurrency` makes all that much sense.